Teach exclusivity access marker verifier to handle UnsafePointer.

atrick opened the pull request:
My previous attempt to strengthen this verification ignored the fact that a
projection or addressor that returns UnsafePointer can have nested access after
being passed as an @inout argument. After inlining, this caused the verifier to
assert.

Instead, handle access to an UnsafePointer as a valid but unidentified access. I was trying
to avoid this because an UnsafePointer could refer to a global or class
property, which we can never consider "Unidentified". However, this is
reasonably safe because it can only result from a _nested_ access. If a global
or class property is being addressed, it must already have its own dynamic
access within the addressor, and that access will be exposed to the
optimizer. If a non-public KeyPath is being addressed, a keypath instruction
must already exist somewhere in the module which is exposed to the optimizer.

Fixes <rdar://problem/41660554> Swift CI (macOS release master, OSS): SIL verification failed: Unknown formal access pattern: storage.

atrick committed:
Teach exclusivity access marker verifier to handle UnsafePointer.

My previous attempt to strengthen this verification ignored the fact that a
projection or addressor that returns UnsafePointer can have nested access after
being passed as an @inout argument. After inlining, this caused the verifier to
assert.

Instead, handle access to an UnsafePointer as a valid but unidentified access. I
was trying to avoid this because an UnsafePointer could refer to a global or
class property, which we can never consider "Unidentified". However, this is
reasonably safe because it can only result from a _nested_ access. If a global
or class property is being addressed, it must already have its own dynamic
access within the addressor, and that access will be exposed to the
optimizer. If a non-public KeyPath is being addressed, a keypath instruction
must already exist somewhere in the module which is exposed to the optimizer.

Fixes <rdar://problem/41660554> Swift CI (macOS release master, OSS): SIL verification failed: Unknown formal access pattern: storage.

atrick commented:
@swift-ci test.

jrose-apple commented on test/SILOptimizer/exclusivity_static_diagnostics.sil:
```diff
@@ -1106,3 +1106,31 @@ bb0(%0 : $UnsafePointer<Int32>, %1 : $*Int32):
+
+// The compiler intrinsic _projectXXXKeyPath returns a tuple of
+// UnsafePointer and Optional AnyObject. After inlining, the unsafe
+// pointer may appear to orinate from anywhere, including a phi.
```
Typo: "orinate"

atrick replied:
Oops thanks. Restarting the tests ;)

gottesmm commented on lib/SIL/MemAccessUtils.cpp:
```diff
@@ -181,6 +181,16 @@ static bool isExternalGlobalAddressor(ApplyInst *AI) {
+  assert(isa<BuiltinRawPointerType>(SEI->getType().getASTType()));
+  auto &C = SEI->getModule().getASTContext();
+  auto *decl = SEI->getStructDecl();
+  return (decl == C.getUnsafeMutablePointerDecl()
```
The return value here doesn't need () around it.

gottesmm commented on test/Interpreter/enforce_exclusive_access.swift:
```diff
@@ -306,7 +306,7 @@ ExclusiveAccessTestSuite.test("SequentialKeyPathWritesDontOverlap") {
   c[keyPath: getF] = 7
   c[keyPath: getF] = 8 // no-trap
-  c[keyPath: getF] += c[keyPath: getF] + 1 // no-trap
+  c[keyPath: getF] += c[keyPath: getF] // no-trap
```
Question. Was this filecheck test changed for a reason.

atrick replied:
Cleanup in a closely related test case.

gottesmm reviewed (commented):
I need to think a bit more about the semantics before giving LGTM. But a few nits and also, can you add an interpreter test? That way we know that we definitely handle the keypath version of the test in perpetuity.

atrick commented:
@swift-ci test.

gottesmm commented:
LGTM

atrick committed:
Tweak exclusivity access marker verification.

Narrow the list of address producers that verification can "peek through"
to avoid trigger an assertion in the previous commit with
-enable-verify-exclusivity.

atrick commented:
@swift-ci smoke test and merge.

atrick commented:
@swift-ci smoke test and merge.

atrick commented:
@swift-ci smoke test and merge.

atrick commented:
@swift-ci smoke test and merge.

atrick commented:
@swift-ci smoke test.
